### PR TITLE
[sources] Restructuring / renaming for source error handling

### PR DESCRIPTION
**Pull request by bkirwi**
This is mostly a renaming PR: we're committing to only sending definite errors through to the downstream error collection, so we rename some error types and a vector or two to reflect that.

Secondarily, we also treat fewer psql errors as definite than we used to: errors that would previously wedge the source will now retry. I think this is fine, but don't feel strongly if y'all hate it.

### Motivation
Known-valuable improvement: #15061.

### Tips for reviewer

See the comment about error classes for the questionable bit.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Treat some Postgres errors as transient instead of fatal in the Postgres source.

### CODE REVIEW

File: src/storage/src/source/postgres.rs [hunk 1]
```diff
@@ -57,61 +57,60 @@ static WAL_LAG_GRACE_PERIOD: Duration = Duration::from_secs(5 * 60); // 5 minute
...
-                            )
+            Some(err) => match err.downcast_ref::<DbError>() {
+                Some(db_err) => {
+                    let class = &db_err.code().code()[0..2];
+                    match class {
+                        // See https://www.postgresql.org/docs/current/errcodes-appendix.html
+                        // for the class definitions.
+
+                        // unknown catalog or schema names
+                        "3D" | "3F" => true,
```
Comment (bkirwi) [inline src/storage/src/source/postgres.rs:80]: This is enough to get the tests to pass, but I'm not totally confident that this is a full enough set of error classes for production. If this seems too risky for now, I can back this part of the PR out... though that of course risks that we treat an error as definite that shouldn't be.

File: src/storage/src/source/postgres.rs [hunk 2]
```diff
@@ -57,61 +57,60 @@ static WAL_LAG_GRACE_PERIOD: Duration = Duration::from_secs(5 * 60); // 5 minute
...
-                            || db_err.code() == &SqlState::CANNOT_CONNECT_NOW
-                            || db_err.code() == &SqlState::ADMIN_SHUTDOWN
-                            || db_err.code() == &SqlState::CRASH_SHUTDOWN
-                            || !matches!(
-                                db_err.parsed_severity(),
-                                Some(Error) | Some(Fatal) | Some(Panic)
-                            )
+            Some(err) => match err.downcast_ref::<DbError>() {
+                Some(db_err) => {
+                    let class = &db_err.code().code()[0..2];
```
Comment (cjubb39) [inline src/storage/src/source/postgres.rs:74]: (definitely deferring to postgres experts on this one)

Why is this approach (extra `.code()[0..2]`) preferable to the previous approach?  The three categories here seem to only have one error in each class?

That being said, if you'd like to keep the generality, what do you think of `let class = db_err.code().code().get(0..2)` where `class` is now an `Option`?  There's nothing that prevent `SqlState::code` from returning a `str` with too few characters -- so we could be defensive here so some garbage from an external system doesn't crash the process

File: src/storage/src/source/source_reader_pipeline.rs [hunk 3]
```diff
@@ -356,17 +354,20 @@ where
                                 }
                             }
                         }
-                        // TODO: Report these errors to the Healthchecker!
                         Some(Err(e)) => {
```
Comment (cjubb39) [inline src/storage/src/source/source_reader_pipeline.rs:357]: to check my understanding: now we only expect definite errors to be returned from the `source_stream` right?

Comment (bkirwi) [reply src/storage/src/source/source_reader_pipeline.rs:357]: Yes, that's right! In the past only `Fatal` errors made it this far.

**Comment by bkirwi** [comment]
I had a somewhat-alarming failure on "checks + kill storaged" just earlier; relevant log:

```> SELECT f1, f2, SUM(LENGTH(f3)) FROM postgres_source_tableA GROUP BY f1, f2;
rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 113ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 22s 33s 49s 74s 78s
<stdin>:2:1: error: non-matching rows: expected:
[["A", "8", "102400"], ["B", "8", "102400"], ["C", "7", "102400"], ["D", "6", "102400"], ["E", "5", "102400"], ["F", "4", "102400"], ["G", "3", "102400"], ["H", "2", "102400"]]
got:
[["A", "3", "102400"], ["A", "4", "-102400"], ["A", "8", "102400"], ["B", "3", "102400"], ["B", "4", "-102400"], ["B", "8", "102400"], ["C", "2", "102400"], ["C", "3", "-102400"], ["C", "7", "102400"], ["D", "2", "-102400"], ["D", "6", "102400"], ["E", "1", "-102400"], ["E", "5", "102400"], ["F", "4", "102400"], ["G", "3", "102400"], ["H", "2", "102400"]]
Poor diff:
+ A 3 102400
+ A 4 -102400
+ B 3 102400
+ B 4 -102400
+ C 2 102400
+ C 3 -102400
+ D 2 -102400
+ E 1 -102400```

This seems new? I'll see if I can reproduce.

**Comment by bkirwi** [comment]
An argument for why the flake can't be related to the changes in this PR:
- Aside from the renaming, the only change is to treat errors that were previously definite as indefinite.
- Only `postgres_replication_loop_inner` checks whether an error is definite or indefinite. All indefinite errors will either panic (if during a snapshot) or log (if during normal operation). In either case, it would appear in the build log of [the failing build](https://buildkite.com/materialize/tests/builds/42709#0183ae10-539a-4b78-bf0d-0d131c2ed03a)... but it does not.

If we're still uncomfortable / the above argument seems too sketchy, we could revert the postgres-error-classification change and keep the renaming. I'll wait for a postgres-source expert to weigh in here I think.

**Comment by cjubb39** [comment]
I _think_ I've seen that error recently.  Your explanation seems reasonable enough to me so I'd guess it's related to work on the source pipeline that @petrosagg et al have been working on.  So all the better reason to wait for postgres review!

(in case not clear, non-postgresy stuff looks good to me!  Thanks for making the `.get(..)` change!)

**Comment by petrosagg** [comment]
So this failure is extremely scary and I cannot find any explanation other than either a bug in the generated dataflow or some even scarier bug in our sink code. I don't think the produced output can be caused by a misbehaving source.

The query of this test is the following: 
```sql
SELECT f1, f2, SUM(LENGTH(f3)) FROM postgres_source_tableA GROUP BY f1, f2;
```

And we see that the actual output of the query in the test was:

```
A, 3, 102400
A, 4, -102400
A, 8, 102400
B, 3, 102400
B, 4, -102400
B, 8, 102400
C, 2, 102400
C, 3, -102400
C, 7, 102400
D, 2, -102400
D, 6, 102400
E, 1, -102400
E, 5, 102400
F, 4, 102400
G, 3, 102400
H, 2, 102400
```

The scary part is that the third column is a summation of the length of a string column, i.e a summation of numbers greater than or equal to zero so the sum should never be negative.

Nevertheless the query did return rows where the sum was negative! How is this possible?

**Comment by frankmcsherry** [comment]
> Nevertheless the query did return rows where the sum was negative! How is this possible?

Internally, operators like `SUM(LENGTH(_))` translate updates like `("hello", t, +1)` to `((), t, +5)` and `("farewell", t, -1)` to `((), t, -8)`. The output you have above *could* result from a bad source that presents a retraction without a corresponding insertion. I don't know that it does, but if at all possible I would try and capture the corresponding input updates to be quite certain. You should also be able to see it with `SELECT f1, f2, COUNT(*) FROM postgres_source_tableA GROUP BY f1, f2;`

**Comment by guswynn** [comment]
Anecdotally, i always ALWAYS seen this error along with this panic in the logs: https://github.com/MaterializeInc/materialize/issues/15046

its unclear to me how a panicking operator can cause us to erroneously close a timestamp with retractions and not insertions. Note that in pg sources, for updates we do send retractions before insertions

@frankmcsherry is it possible that a panic in an operator can drop a `CapabilitySet` and that progress can be communicated through the rest of the dataflow graph before we abort the process? To me that seems impossible, especially for a single-worker setup with a single timely thread

**Comment by frankmcsherry** [comment]
> is it possible that ...

I can never tell what happens with `async`, but not only would the `CapabilitySet` need to be dropped, we would also need to continue to schedule the containing `Scope` that is in charge of reading out and disseminating the progress information. It seems like there would need to be a panic handler somewhere for this to happen (`async`?) but timely does not introduce one.

**Comment by guswynn** [comment]
@frankmcsherry additionally, this appears to be happening with a single timely thread, so that thread panicking (async or not, we aren't running the operators in tokio tasks that catch panics) would prevent any other progress...which makes this even harder to figure out

**Comment by philip-stoev** [comment]
The negative result from the `SUM` now has its own ticket:
 - #15350

**Comment by bkirwi** [comment]
Cool! Now that we're not worried that this PR caused our mysterious postgres error, I've updated against the latest main.

@petrosagg / @aljoscha - this seem reasonable / consistent with what we talked about earlier this week?

File: src/storage/src/source/postgres.rs [hunk 4]
```diff
@@ -57,78 +57,80 @@ static WAL_LAG_GRACE_PERIOD: Duration = Duration::from_secs(5 * 60); // 5 minute
...
+    fn is_definite(&self) -> bool;
 }
 
 impl ErrorExt for tokio::time::error::Elapsed {
-    fn is_recoverable(&self) -> bool {
-        true
+    fn is_definite(&self) -> bool {
+        false
     }
 }
```
Comment (guswynn) [inline src/storage/src/source/postgres.rs:67]: what code breaks if we remove this impl?

File: src/storage/src/source/postgres.rs [hunk 5]
```diff
@@ -57,78 +57,80 @@ static WAL_LAG_GRACE_PERIOD: Duration = Duration::from_secs(5 * 60); // 5 minute
...
-                            || !matches!(
-                                db_err.parsed_severity(),
-                                Some(Error) | Some(Fatal) | Some(Panic)
-                            )
+            Some(err) => match err.downcast_ref::<DbError>() {
+                Some(db_err) => {
+                    let class = match db_err.code().code().get(0..2) {
+                        None => return false,
+                        Some(class) => class,
+                    };
```
Comment (guswynn) [inline src/storage/src/source/postgres.rs:77]: til about pg error classes

File: src/storage/src/source/postgres.rs [hunk 6]
```diff
@@ -405,7 +407,7 @@ async fn postgres_replication_loop_inner(
                     &task_info.source_id
                 );
             }
-            Err(ReplicationError::Recoverable(e)) => {
+            Err(ReplicationError::Indefinite(e)) => {
```
Comment (guswynn) [inline src/storage/src/source/postgres.rs:410]: i wouldn't describe this as `Indefinite`, as it never errors the source (same with the retrying in the replication stream below); what if we made the 2 variants `Retryable` and `Definite`?

File: src/storage/src/source/postgres.rs (hunk 4 above)
Comment (bkirwi) [reply src/storage/src/source/postgres.rs:67]: A tokio timeout! https://github.com/MaterializeInc/materialize/blob/a9b3e50d95e37c1fe18973a94bfc2c7069b292a9/src/storage/src/source/postgres.rs#L673-L674

File: src/storage/src/source/postgres.rs (hunk 5 above)
Comment (bkirwi) [reply src/storage/src/source/postgres.rs:77]: Yeah same TBH! Seemed less error prone than listing out the errors explicitly.

File: src/storage/src/source/postgres.rs (hunk 6 above)
Comment (bkirwi) [reply src/storage/src/source/postgres.rs:410]: No objection! Curious: why does "indefinite" imply that it should error the source?

Comment (cjubb39) [reply src/storage/src/source/postgres.rs:410]: yeah +1.  I like sticking with one set of semantics here: definite vs indefinite/nondefinite seems to be wording we use throughout the code?  Unless, of course, there's a reason missing that we shouldn't!

**Pull request merged**